# CRISP DM


## Problem Understanding:
A central step in the ORCA pipeline is the automated mapping of natural language threat
descriptions to structured attack patterns from the Common Attack Pattern Enumeration
and Classification (CAPEC) framework. This process—referred to as Threat-to-CAPEC
Mapping—aims to translate unstructured textual threat inputs into standardized, machine-
readable representations that describe how an attacker might exploit a given vulnerability.
This mapping is critical for enabling further steps in the security analysis process, such
as correlating threats with known vulnerabilities (CWEs, CVEs), assessing risk using
scoring systems like CVSS, and informing mitigation strategies. However, the task poses
several inherent challenges:
Ambiguity: Threat descriptions are often informal, incomplete, or context-dependent.
Terminology mismatch: Natural language inputs may not directly align with the technical
vocabulary used in CAPEC definitions.
Granularity: A single threat may correspond to multiple CAPECs at varying levels of
abstraction, requiring semantic reasoning to determine relevance.
Solving this problem involves designing a system capable of understanding the semantics
of threat descriptions and reliably identifying the most appropriate CAPEC entries. This
step must balance accuracy, scalability, and interpretability to support reliable, automated
security assessments within the broader ORCA framework.

## Data Understanding

In [21]:
# imports
import pandas as pd
import json

In [18]:
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

Import threat data from json file

In [5]:
# Load threat data from JSON file into a DataFrame
df_threats = pd.read_json('data/all_threats.json')
df_threats.head()

Threat ID                                       Threat title  \
0  T-O-RAN-01  An attacker exploits insecure designs or lack ...   
1  T-O-RAN-02  An attacker exploits misconfigured or poorly c...   
2  T-O-RAN-03  Attacks from the internet to penetrate O-RAN n...   
3  T-O-RAN-04  An attacker attempts to jam the airlink signal...   
4  T-O-RAN-05  An attacker penetrates and compromises the O-R...   

                                  Threat Description Threat type Impact type  \
0  Unauthenticated/unauthorized access to O-RAN c...        None        None   
1  Unauthenticated/unauthorized access to O-RAN c...        None        None   
2  Web servers serving O-RAN functional and manag...        None        None   
3  DDoS attacks on O-RAN systems: The 5G evolutio...        None        None   
4  O-RAN’s Fronthaul, O1, O2, A1, and E2 manageme...        None        None   

  Threat agent                                      Vulnerability  \
0          All  [Outdated component from the lack of update or...   
1          All  [Errors from the lack of configuration change ...   
2          All  [Errors in the design and implementation of th...   
3          All           [Failure to address overload situations]   
4          All  [Improper or missing authentication and author...   

         Threatened Asset                                Affected Components  
0                     All                                                All  
1                     All                                                All  
2                     All                                                All  
3  ASSET-D-06, ASSET-D-18                        O-RU, airlink with UE, O-DU  
4                     All  rApps, xApps, O-RU, O-DU, O-CU, Near-RT RIC, N...

In [ ]:
# Display DataFrame information
df_threats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 182 entries, 0 to 181
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Threat ID            182 non-null    object
 1   Threat title         182 non-null    object
 2   Threat Description   182 non-null    object
 3   Threat type          82 non-null     object
 4   Impact type          82 non-null     object
 5   Threat agent         182 non-null    object
 6   Vulnerability        182 non-null    object
 7   Threatened Asset     182 non-null    object
 8   Affected Components  182 non-null    object
dtypes: object(9)
memory usage: 12.9+ KB


There a 182 threats in 9 columns, each threat has a unique ID, a title, a description, a threat type, an impact type, a threat agent, one or more vulnerabilities (safed in a list), a threatened asset and a affected component

In [12]:
df_threats.describe()

Threat ID                                       Threat title  \
count          182                                                182   
unique         182                                                181   
top     T-O-RAN-01  External attacker exploits authentication weak...   
freq             1                                                  2   

                                       Threat Description Threat type  \
count                                                 182          82   
unique                                                182          16   
top     Unauthenticated/unauthorized access to O-RAN c...    Spoofing   
freq                                                    1          28   

         Impact type Threat agent                 Vulnerability  \
count             82          182                           182   
unique            17            2                           116   
top     Authenticity          All  [weak mutual authentication]   
freq              26          178                            12   

                                         Threatened Asset Affected Components  
count                                                 182                 182  
unique                                                 58                  55  
top     ASSET-D-12, ASSET-D-13, ASSET-D-14, ASSET-D-15...                 All  
freq                                                   24                  19

In [26]:
df_threats.iloc[0]

Threat ID                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

In [22]:
print(json.dumps(df_threats.iloc[0].to_dict(), indent=2, ensure_ascii=False))

{
  "Threat ID": "T-O-RAN-01",
  "Threat title": "An attacker exploits insecure designs or lack of adaption in O-RAN components",
  "Threat Description": "Unauthenticated/unauthorized access to O-RAN components could possibly be achieved via the different O-RAN interfaces, depending upon the design of the hardware-software O-RAN system and how different functions are segregated within the O-RAN system. \nO-RAN components might be vulnerable if: \n• Outdated component from the lack of update or patch management,\n• Poorly design architecture,\n• Missing appropriate security hardening,\n• Unnecessary or insecure function/protocol/component.\nAn attacker could, in such case, either inject malwares and/or manipulate existing software, harm the O-RAN components, create a performance issue by manipulation of parameters, or reconfigure the O-RAN components and disable the security features with the purpose of eavesdropping or wiretapping on various CUS & M planes, reaching northbound systems,

In [28]:
from IPython.display import HTML, display

display(HTML('''
<style>
/* White background for output area */
.output_area {
    background: white !important;
    color: black !important;
}

/* Optional: White background for DataFrame cells */
.dataframe {
    background-color: white !important;
    color: black !important;
}
</style>
'''))



In [29]:
display(df_threats)

Threat ID  \
0    T-O-RAN-01   
1    T-O-RAN-02   
2    T-O-RAN-03   
3    T-O-RAN-04   
4    T-O-RAN-05   
..          ...   
177     T-E2-02   
178     T-E3-03   
179     T-Y1-01   
180     T-Y1-02   
181     T-Y1-03   

                                                                                                       Threat title  \
0                                     An attacker exploits insecure designs or lack of adaption in O-RAN components   
1                                          An attacker exploits misconfigured or poorly configured O-RAN components   
2                                                     Attacks from the internet to penetrate O-RAN network boundary   
3                                                An attacker attempts to jam the airlink signal through IoT devices   
4    An attacker penetrates and compromises the O-RAN system through the open O-RAN’s Fronthaul, O1, O2, A1, and E2   
..                                                                                                              ...   
177                                                          Malicious actor monitors messaging across E2 interface   
178                                                          Malicious actor modifies messaging across E2 interface   
179                                                                          Untrusted Near-RT-RIC and Y1 consumers   
180                                                          Malicious actor monitors messaging across Y1 interface   
181                                                          Malicious actor modifies messaging across Y1 interface   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Threat Description  \
0    Unauthenticated/unauthorized access to O-RAN components could possibly be achieved via the different O-RAN interfaces, depending upon the design of the hardware-software O-RAN system and how different functions are segregated within the O-RAN system. \nO-RAN components might be vulnerable if: \n• Outdated component from the lack of update or patch management,\n• Poorly design architecture,\n• Missing appropriate security hardening,\n• Unnecessary or insecure function/protocol/component.\nAn attacker could, in such case, either inject malwares and/or manipulate existing software, harm the O-RAN components, create a performance issue by manipulation of parameters, or reconfigure the O-RAN components and disable the security features with the purpose of eavesdropping or wiretapping on various CUS & M planes, reaching northbound systems, attack broader network to cause denial-of-service, steal unprotected private keys, certificates, hash values, or other type of breaches.\nIn addition, O-RAN components could be software providing network functions, so they are likely to be vu